### PR TITLE
Parallelize clustering

### LINKED ISSUE #18: Parallelizing the clustering step to improve efficiency
Platform: Linux, Windows
Version: Latest

Currently, during sampling exhaustiveness checking, the clustering of the good-scoring-models at various cutoff thresholds happens sequentially under a `for` loop which takes quite a bit of time given a large number of models and a large number of thresholds to check for. To improve the time-efficiency of the sampcon pipeline, can this step be parallelized since each clustering is independent?

PS: In case a simple `multiprocessing` based parallelization is enough, I am opening a PR to achieve that. If this is sufficient, perhaps the initial steps of parsing and reading the RMF files into memory can be parallelized in a similar way (can open a PR for that too).

### PR DESCRIPTION
**Pull request by Stochastic13**
Added multiprocessing based parallelization of the clustering step
Fixes #18

### CODE REVIEW

**Comment by Stochastic13** [comment]
The `import os` statement is only to be able to use `os.cpu_count()` for parallelization downstream in the file, but since the `exhaust.py` invocation collects the preferred number of cores as the argument `-c`, the `import` statement might be redundant.

**Comment by shruthivis** [comment]
@benmwebb can you review and merge? This is from Satwik of my lab and I've already reviewed the code.

**Comment by benmwebb** [comment]
> @benmwebb can you review and merge? This is from Satwik of my lab and I've already reviewed the code.

Happy to look at it when I'm back from vacation next week. In the meantime, you should take a look to see why the tests are failing. They need to work to merge the code of course.

**Comment by Stochastic13** [comment]
@benmwebb  

The `python 2.7` version tests are failing since the `multiprocessing.starmap` was included in `python 3.3+`. I can replace this with the `2.7`-compatible `multiprocessing.map` with a few tweaks in the argument passing of the function. 

I am not sure why the `python 3.9` tests fail since the output show that all the 22 tests passed with an unrelated warning.

**Comment by benmwebb** [comment]
> I am not sure why the `python 3.9` tests fail since the output show that all the 22 tests passed with an unrelated warning.

The errors are from `flake8`:

```
2021-09-14T17:21:58.0737232Z ../pyext/src/clustering_rmsd.py:6:1: F401 'os' imported but unused
2021-09-14T17:21:58.0739140Z ../pyext/src/clustering_rmsd.py:151:80: E501 line too long (111 > 79 characters)
2021-09-14T17:21:58.0740487Z ../pyext/src/exhaust.py:273:80: E501 line too long (81 > 79 characters)
```

**Comment by Stochastic13** [comment]
@benmwebb My bad. I ran my local code-formatting setup throughout the two files and hopefully they are all `PEP8` compatible now. Fixed the python2.7 compatibility issue.

File: pyext/src/clustering_rmsd.py [hunk 1]
```diff
@@ -139,18 +139,25 @@ def percent_ensemble_explained(ctable, total_num_models):
...
     percents = []
+
+    def unpacking_wrapper(arg_tuple):
+        return precision_cluster(*arg_tuple)
+
     with open("%s.ChiSquare_Grid_Stats.txt" % sysname, 'w+') as f1:
-        for c in cutoffs_list:
-            cluster_centers, cluster_members = precision_cluster(
-                    distmat_full, total_num_models, c)
+        with Pool(cores) as p:
```
Comment (benmwebb) [inline pyext/src/clustering_rmsd.py:152]: I don't think this will work in Python 2 because Pool doesn't support the context manager protocol until 3.4 or so. IMP code still needs to work with both Python 3 and Python 2.

File: pyext/src/clustering_rmsd.py [hunk 2]
```diff
@@ -23,12 +24,11 @@ def get_sample_identity(idfile_A, idfile_B):
 
 
 def get_cutoffs_list(distmat, gridSize):
-
     mindist = distmat.min()
     maxdist = distmat.max()
 
     print("Minimum and maximum pairwise model distances:", mindist, maxdist)
-    cutoffs = numpy.arange(mindist+gridSize, maxdist, gridSize)
+    cutoffs = numpy.arange(mindist + gridSize, maxdist, gridSize)
```
Comment (benmwebb) [inline pyext/src/clustering_rmsd.py:31]: Most of the changes in this PR appear to be minor formatting changes like this one. This makes it quite hard to spot what it is you actually changed. I'm certainly not opposed to making the code easier to read if you really want to, but it should be in a separate PR.

File: pyext/src/exhaust.py [hunk 3]
```diff
@@ -269,14 +269,15 @@ def main():
 
         # Do clustering at each cutoff
         pvals, cvs, percents = clustering_rmsd.get_clusters(
-                cutoffs_list, distmat_full, all_models, total_num_models,
-                sampleA_all_models, sampleB_all_models, args.sysname)
+            cutoffs_list, distmat_full, all_models, total_num_models,
+            sampleA_all_models, sampleB_all_models, args.sysname,
+            args.cores)
```
Comment (benmwebb) [inline pyext/src/exhaust.py:274]: Looks like you are extending the `cores` command line argument here, so you should change the help text for it too. Currently it says it's only affects the RMSD matrix calculations with cpu_omp; that's no longer the case.

**Review by benmwebb — CHANGES_REQUESTED** [review]
Looks mostly OK to me. Strip out the random whitespace changes so the PR just contains the parallelization change... and ensure the tests pass (they currently don't), and I'll be happy to merge it.

File: pyext/src/clustering_rmsd.py (hunk 1 above)
Comment (Stochastic13) [reply pyext/src/clustering_rmsd.py]: I changed it to not using the context manager. Hopefully, this should run on Python 2 also. In case it doesn't I can run it without `Pool` (using `multiprocessing.Process`)
